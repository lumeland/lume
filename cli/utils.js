import { existsSync } from "../deps/fs.ts";
import lume, { overrides } from "../mod.js";
import { join, resolve, toFileUrl } from "../deps/path.ts";
import { bold, brightGreen, dim, red } from "../deps/colors.ts";

/**
 * Create and configure a site instance
 *
 * @param {Object} options user options to override
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  overrides.cwd = options.root;

  if (options.dev) {
    overrides.dev = options.dev;
  }

  if (options.location) {
    overrides.location = new URL(options.location);
  }

  if (options.src) {
    overrides.src = options.src;
  }

  if (options.dest) {
    overrides.dest = options.dest;
  }

  if (options.port) {
    (overrides.server ||= {}).port = parseInt(options.port);
  }

  if (options.open) {
    (overrides.server ||= {}).open = options.open;
  }

  if (options["--"]) {
    overrides.flags = options["--"];
  }

  if (existsSync(configFile)) {
    const mod = await import(toFileUrl(configFile));
    return mod.default;
  }

  return lume();
}

/**
 * Validate the number of arguments passed to a command
 *
 * @param {string} command name of the command you're validating
 * @param {[]} args array of cli arguments (i.e. options._)
 * @param {number} max the max number of args
 * @param {number} min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}

export function getCurrentVersion() {
  const url = new URL("../", import.meta.url).pathname;
  return url.match(/@([^/]+)/)?.[1] ?? `local (${url})`;
}

export function printError(exception, debug = false) {
  console.log();
  console.error(bold(red(`Error:`)), exception.message);

  printDataError(exception);

  if (debug) {
    console.error(exception);
  }

  if (exception.error) {
    let { error } = exception;
    let indent = "  ";
    while (error) {
      console.log();
      console.log(red(`${indent}${error.message}`));
      printDataError(error, indent);
      indent += "  ";

      if (debug) {
        console.error(error);
      }

      error = error.error;
    }
  }

  console.log();
}

function printDataError(error, indent = "") {
  if (!error.data) {
    return;
  }

  for (let [key, value] of Object.entries(error.data)) {
    if (key === "page") {
      value = value.src.path + value.src.ext;
    }

    console.log(dim(`${indent}${key}:`), value);
  }
}
