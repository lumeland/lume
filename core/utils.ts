import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import { SEP } from "../deps/path.ts";

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

/**
 * The list of supported MIME types.
 * It's used by the server and some plugins.
 */
export const mimes: Map<string, string> = new Map([
  [".apng", "image/apng"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".avif", "image/avif"],
  [".css", "text/css; charset=utf-8"],
  [".gif", "image/gif"],
  [".html", "text/html; charset=utf-8"],
  [".ico", "image/x-icon"],
  [".jpeg", "image/jpg"],
  [".jpg", "image/jpg"],
  [".js", "text/javascript; charset=utf-8"],
  [".json", "application/json"],
  [".map", "application/json"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mp4"],
  [".ogg", "audio/ogg"],
  [".otf", "font/otf"],
  [".pdf", "application/pdf"],
  [".png", "image/png"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".ttf", "font/ttf"],
  [".txt", "text/plain; charset=utf-8"],
  [".wasm", "application/wasm"],
  [".webm", "video/webm"],
  [".webmanifest", "application/manifest+json"],
  [".webp", "image/webp"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".xml", "text/xml"],
  [".zip", "application/zip"],
]);

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type>,
) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

/** Check if the argument passed is a plain object */
function isPlainObject(obj: unknown) {
  return typeof obj === "object" && obj !== null &&
    obj.toString() === "[object Object]";
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/").
 */
export function normalizePath(path: string) {
  return SEP === "/" ? path : path.replaceAll(SEP, "/");
}

/**
 * Search an extension in a map.
 * It's useful for cases in which the extension is multiple.
 * Example: page.tmpl.js
 */
export function searchByExtension<Type>(
  path: string,
  extensions: Map<string, Type>,
): [string, Type] | undefined {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return documentElement?.outerHTML || "";
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/**
 * Generic Exception to throw errors.
 * It allows to include extra data and the previous exception.
 */
export class Exception extends Error {
  data?: Record<string, unknown>;
  error?: Error;

  constructor(message: string, data?: Record<string, unknown>, error?: Error) {
    super(message);
    this.data = data;
    this.error = error;
  }
}
