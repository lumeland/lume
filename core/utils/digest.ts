import { crypto } from "../../deps/crypto.ts";
import { encodeHex } from "../../deps/hex.ts";

const decoder = new TextDecoder();
const encoder = new TextEncoder();

/** Digest a message using SHA-1 algorithm */
export async function sha1(
  message: string | Uint8Array<ArrayBuffer>,
): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("SHA-1", message);
  return decoder.decode(hash);
}

/** Digest a message using MD5 algorithm */
export async function md5(
  message: string | Uint8Array<ArrayBuffer>,
): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("MD5", message);
  return encodeHex(hash);
}
