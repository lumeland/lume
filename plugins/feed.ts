import { getExtension } from "../core/utils/path.ts";
import { isPlainObject, merge } from "../core/utils/object.ts";
import { getCurrentVersion } from "../core/utils/lume_version.ts";
import { getDataValue, getPlainDataValue } from "../core/utils/data_values.ts";
import { cdata, stringify } from "../deps/xml.ts";
import { Page } from "../core/file.ts";
import { parseDate } from "../core/utils/date.ts";

import type Site from "../core/site.ts";
import type { Data } from "../core/file.ts";

export interface Options {
  /** The output filenames */
  output?: string | string[];

  /** The query to search the pages */
  query?: string;

  /** The sort order */
  sort?: string;

  /** The maximum number of items */
  limit?: number;

  /** The feed info */
  info?: FeedInfoOptions;

  /** The feed items configuration */
  items?: FeedItemOptions;
}

export interface FeedInfoOptions {
  /** The feed title */
  title?: string;

  /** The feed subtitle */
  subtitle?: string;

  /**
   * The feed published date
   * @default `new Date()`
   */
  published?: Date;

  /** The feed description */
  description?: string;

  /** The feed language */
  lang?: string;

  /** The feed generator. Set `true` to generate automatically */
  generator?: string | boolean;

  /** The feed author name */
  authorName?: string;

  /** The feed author URL */
  authorUrl?: string;

  /** The main image of the site */
  image?: string;

  /** The logotype or icon of the site */
  icon?: string;

  /** The color theme of the site */
  color?: string;
}

export interface FeedItemOptions {
  /** The item title */
  title?: string | ((data: Data) => string | undefined);

  /** The item description */
  description?: string | ((data: Data) => string | undefined);

  /** The item published date */
  published?: string | ((data: Data) => Date | undefined);

  /** The item updated date */
  updated?: string | ((data: Data) => Date | undefined);

  /** The item content */
  content?: string | ((data: Data) => string | undefined);

  /** The item language */
  lang?: string | ((data: Data) => string | undefined);

  /** The item image */
  image?: string | ((data: Data) => string | undefined);

  /** The item author name */
  authorName?: string;

  /** The item author URL */
  authorUrl?: string;
}

export const defaults: Options = {
  /** The output filenames */
  output: "/feed.rss",

  /** The query to search the pages */
  query: "",

  /** The sort order */
  sort: "date=desc",

  /** The maximum number of items */
  limit: 10,

  /** The feed info */
  info: {
    title: "My RSS Feed",
    published: new Date(),
    description: "",
    lang: "en",
    generator: true,
  },
  items: {
    title: "=title",
    description: "=description",
    published: "=date",
    content: "=children",
    lang: "=lang",
  },
};

export interface Author {
  name?: string;
  url?: string;
}

export interface FeedData {
  title: string;
  url: string;
  description: string;
  published: Date;
  lang: string;
  generator?: string;
  items: FeedItem[];
  author?: Author;
  image?: string;
  icon?: string;
  color?: string;
}

export interface FeedItem {
  title: string;
  url: string;
  description: string;
  published: Date;
  updated?: Date;
  content: string;
  lang: string;
  image?: string;
  author?: Author;
}

const defaultGenerator = `Lume ${getCurrentVersion()}`;

/**
 * A plugin to generate RSS and JSON feeds
 * @see https://lume.land/plugins/feed/
 */
export function feed(userOptions?: Options) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    site.addEventListener("beforeSave", () => {
      const output = Array.isArray(options.output)
        ? options.output
        : [options.output];

      const pages = site.search.pages(
        options.query,
        options.sort,
        options.limit,
      ) as Data[];

      const { info, items } = options;
      const rootData = site.source.data.get("/") || {};

      const feed: FeedData = {
        title: getPlainDataValue(rootData, info.title),
        description: getPlainDataValue(rootData, info.description),
        published: getDataValue(rootData, info.published),
        lang: getDataValue(rootData, info.lang),
        url: site.url("", true),
        generator: info.generator === true
          ? defaultGenerator
          : info.generator || undefined,
        author: getAuthor(rootData, info),
        image: info.image,
        icon: info.icon,
        color: info.color,
        items: pages.map((data): FeedItem => {
          const content = getDataValue(data, items.content)?.toString();
          const pageUrl = site.url(data.url, true);
          const fixedContent = fixUrls(new URL(pageUrl), content || "");
          const imagePath = getDataValue(data, items.image);
          const image = imagePath !== undefined
            ? site.url(imagePath, true)
            : undefined;

          return {
            title: getPlainDataValue(data, items.title),
            url: site.url(data.url, true),
            description: getPlainDataValue(data, items.description),
            author: getAuthor(data, items),
            published: toDate(getDataValue(data, items.published)) ||
              new Date(),
            updated: toDate(getDataValue(data, items.updated)),
            content: fixedContent,
            lang: getDataValue(data, items.lang),
            image,
          };
        }),
      };

      for (const filename of output) {
        const format = getExtension(filename).slice(1);
        const file = site.url(filename, true);

        switch (format) {
          case "rss":
          case "feed":
          case "xml":
            site.pages.push(
              Page.create({ url: filename, content: generateRss(feed, file) }),
            );
            break;

          case "json":
            site.pages.push(
              Page.create({ url: filename, content: generateJson(feed, file) }),
            );
            break;

          default:
            throw new Error(`Invalid Feed format "${format}"`);
        }
      }
    });
  };
}

function getAuthor(
  data: Partial<Data>,
  info: FeedInfoOptions | FeedItemOptions,
): Author | undefined {
  const name = getPlainDataValue(data, info.authorName);
  const url = getDataValue(data, info.authorUrl);

  if (name || url) {
    return { name, url };
  }
}

function fixUrls(base: URL, html: string): string {
  return html.replaceAll(
    /\s(href|src)="([^"]+)"/g,
    (_match, attr, value) => ` ${attr}="${new URL(value, base).href}"`,
  );
}

function generateRss(data: FeedData, file: string): string {
  const feed = {
    "@version": "1.0",
    "@encoding": "UTF-8",
    rss: {
      "@xmlns:content": "http://purl.org/rss/1.0/modules/content/",
      "@xmlns:wfw": "http://wellformedweb.org/CommentAPI/",
      "@xmlns:dc": "http://purl.org/dc/elements/1.1/",
      "@xmlns:atom": "http://www.w3.org/2005/Atom",
      "@xmlns:sy": "http://purl.org/rss/1.0/modules/syndication/",
      "@xmlns:slash": "http://purl.org/rss/1.0/modules/slash/",
      "@xmlns:webfeeds": "http://webfeeds.org/rss/1.0",
      "@version": "2.0",
      channel: {
        title: data.title,
        link: data.url,
        "atom:link": {
          "@href": file,
          "@rel": "self",
          "@type": "application/rss+xml",
        },
        description: data.description,
        lastBuildDate: data.published.toUTCString(),
        language: data.lang,
        generator: data.generator,
        author: {
          name: data.author?.name,
          uri: data.author?.url,
        },
        "webfeeds:cover": {
          "@image": data.image,
        },
        "webfeeds:logo": data.icon,
        "webfeeds:accentColor": data.color,
        item: data.items.map((item) => ({
          title: item.title,
          link: item.url,
          guid: {
            "@isPermaLink": false,
            "#text": item.url,
          },
          author: {
            name: item.author?.name,
            uri: item.author?.url,
          },
          description: item.description,
          "content:encoded": cdata(item.content),
          pubDate: item.published.toUTCString(),
          "atom:updated": item.updated?.toISOString(),
          meta: item.image
            ? { "@property": "og:image", "@content": item.image }
            : undefined,
        })),
      },
    },
  };

  return stringify(clean(feed));
}

function generateJson(data: FeedData, file: string): string {
  const feed = {
    version: "https://jsonfeed.org/version/1",
    title: data.title,
    home_page_url: data.url,
    feed_url: file,
    description: data.description,
    author: data.author,
    icon: data.image,
    favicon: data.icon,
    items: data.items.map((item) => ({
      id: item.url,
      url: item.url,
      title: item.title,
      author: item.author,
      content_html: item.content,
      date_published: item.published.toUTCString(),
      date_modified: item.updated?.toUTCString(),
      image: item.image,
    })),
  };

  return JSON.stringify(clean(feed));
}

/** Remove undefined values of an object recursively */
function clean(obj: Record<string, unknown>): Record<string, unknown> {
  return Object.fromEntries(
    Object.entries(obj)
      .map(([key, value]): [string, unknown] => {
        if (isPlainObject(value)) {
          const cleanValue = clean(value);
          return [
            key,
            Object.keys(cleanValue).length > 0 ? cleanValue : undefined,
          ];
        }
        if (Array.isArray(value)) {
          const cleanValue = value
            .map((v) => isPlainObject(v) ? clean(v) : v)
            .filter((v) => v !== undefined);
          return [
            key,
            cleanValue.length > 0 ? cleanValue : undefined,
          ];
        }
        return [key, value];
      })
      .filter(([, value]) => value !== undefined),
  );
}

function toDate(date?: string | number | Date): Date | undefined {
  if (date instanceof Date) {
    return date;
  }
  if (date === undefined) {
    return;
  }
  return parseDate(date);
}

export default feed;
