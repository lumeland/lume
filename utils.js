import { DOMParser } from "./deps/dom.ts";
import { SEP } from "./deps/path.ts";

export async function concurrent(iterable, iteratorFn, limit = 200) {
  const executing = [];

  for await (const item of iterable) {
    const p = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

export const mimes = new Map([
  [".apng", "image/apng"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".avif", "image/avif"],
  [".css", "text/css; charset=utf-8"],
  [".gif", "image/gif"],
  [".html", "text/html; charset=utf-8"],
  [".ico", "image/x-icon"],
  [".jpeg", "image/jpg"],
  [".jpg", "image/jpg"],
  [".js", "text/javascript; charset=utf-8"],
  [".json", "application/json"],
  [".map", "application/json"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mp4"],
  [".ogg", "audio/ogg"],
  [".otf", "font/otf"],
  [".pdf", "application/pdf"],
  [".png", "image/png"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".ttf", "font/ttf"],
  [".txt", "text/plain; charset=utf-8"],
  [".wasm", "application/wasm"],
  [".webm", "video/webm"],
  [".webmanifest", "application/manifest+json"],
  [".webp", "image/webp"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".xml", "text/xml"],
  [".zip", "application/zip"],
]);

export function merge(defaults, user) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      merged[key] = merge(merged[key], value);
      continue;
    }

    merged[key] = value;
  }

  return merged;
}

function isPlainObject(obj) {
  return typeof obj === "object" && obj.toString() === "[object Object]";
}

export function normalizePath(path) {
  return SEP === "/" ? path : path.replaceAll(SEP, "/");
}

export function searchByExtension(path, extensions) {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}

export function documentToString(document) {
  const { doctype } = document;

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${document.documentElement.outerHTML}`;
}

const parser = new DOMParser();

export function stringToDocument(string) {
  return parser.parseFromString(string, "text/html");
}

export class Exception extends Error {
  constructor(message, data, error) {
    super(message);
    this.data = data;
    this.error = error;
  }
}
