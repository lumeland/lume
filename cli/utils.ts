import { Page, Site } from "../core.ts";
import lume from "../mod.ts";
import { exists } from "../deps/fs.ts";
import { join, resolve, toFileUrl } from "../deps/path.ts";
import { bold, dim, red } from "../deps/colors.ts";
import { Exception } from "../core/utils.ts";

/** Return the current installed version */
export function getCurrentVersion(): string {
  const url = new URL("../", import.meta.url).pathname;
  return url.match(/@([^/]+)/)?.[1] ?? `local (${url})`;
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLastVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLastDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    "https://api.github.com/repos/lumeland/lume/commits?per_page=1",
  );
  const commits = await response.json();
  return commits[0].sha;
}

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  root = resolve(Deno.cwd(), root);

  if (config) {
    const path = join(root, config);

    if (await exists(path)) {
      console.log(`Loading config file ${dim(config)}`);
      console.log();
      const mod = await import(toFileUrl(path).href);
      return mod.default;
    }

    throw new Exception("Config file not found", { path });
  }

  const files = ["_config.js", "_config.ts"];

  for (const file of files) {
    const path = join(root, file);

    if (await exists(path)) {
      console.log(`Loading config file ${dim(file)}`);
      console.log();
      const mod = await import(toFileUrl(path).href);
      return mod.default;
    }
  }

  return lume();
}

/** Pretty-print an Error or Exception instance */
export function printError(
  error: Error | Exception,
  indent = 0,
  stackLines = 1,
) {
  console.log();
  const tab = "  ".repeat(indent);

  console.error(`${tab}${bold(red(`Error:`))}`, error.message);

  if (error instanceof Exception) {
    for (let [key, value] of Object.entries(error.data ?? {})) {
      if (key === "page") {
        value = (value as Page).src.path + (value as Page).src.ext;
      }
      console.log(dim(`${tab}${key}:`), value);
    }
  }

  if (error.stack) {
    const stack = error.stack.split("\n");

    // Skip all the stack lines that have been already presented
    stack.slice(1, stack.length - stackLines).forEach((line) => {
      console.log(`${tab}${line.trim()}`);
      stackLines++;
    });
  }

  if (error instanceof Exception && error.error) {
    printError(error.error, indent + 1, stackLines);
  }

  if (indent == 0) {
    console.log();
  }
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "bundler",
  "code_highlight",
  "date",
  "eta",
  "inline",
  "jsx",
  "postcss",
  "pug",
  "relative_urls",
  "slugify_urls",
  "svg",
  "terser",
];
