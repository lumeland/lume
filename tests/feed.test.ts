import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import feed from "../plugins/feed.ts";

Deno.test("RSS plugin", async (t) => {
  const site = getSite({
    src: "feed",
    location: new URL("https://example.com/"),
  });

  site.use(
    feed({
      output: ["feed.json", "feed.rss"],
      info: {
        published: new Date("2020-01-01"),
        generator: "https://lume.land",
        authorName: "Laura Rubio",
        icon: "https://example.com/icon.svg",
        image: "https://example.com/image.png",
        color: "#ff0000",
      },
      items: {
        title: (data) => data.title?.toUpperCase(),
        updated: "=date || $.date",
        published: "$.not-found || =published",
        authorName: "=author.name",
        authorUrl: "=author.url",
      },
    }),
  );
  site.ignore("static.yml");

  // build page
  await build(site);

  // compare to snapshot
  await assertSiteSnapshot(t, site);
});
