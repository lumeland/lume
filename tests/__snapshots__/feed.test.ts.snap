export const snapshot = {};

snapshot[`RSS plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/404.md",
    "/page1.md",
    "/page5.yaml",
    "/static.yml",
  ],
}
`;

snapshot[`RSS plugin 2`] = `[]`;

snapshot[`RSS plugin 3`] = `
[
  {
    content: '{"version":"https://jsonfeed.org/version/1","title":"My RSS Feed","home_page_url":"https://example.com/","feed_url":"https://example.com/feed.json","author":{"name":"Laura Rubio"},"icon":"https://example.com/image.png","favicon":"https://example.com/icon.svg","items":[{"id":"https://example.com/page5/","url":"https://example.com/page5/","title":"PAGE 5","author":{"name":"Óscar","url":"https://oscarotero.com"},"content_html":"Content of Page 5","date_published":"Thu, 21 Jun 1979 23:45:00 GMT","date_modified":"Thu, 21 Jun 1979 23:45:00 GMT"}]}',
    data: {
      basename: "feed",
      content: '{"version":"https://jsonfeed.org/version/1","title":"My RSS Feed","home_page_url":"https://example.com/","feed_url":"https://example.com/feed.json","author":{"name":"Laura Rubio"},"icon":"https://example.com/image.png","favicon":"https://example.com/icon.svg","items":[{"id":"https://example.com/page5/","url":"https://example.com/page5/","title":"PAGE 5","author":{"name":"Óscar","url":"https://oscarotero.com"},"content_html":"Content of Page 5","date_published":"Thu, 21 Jun 1979 23:45:00 GMT","date_modified":"Thu, 21 Jun 1979 23:45:00 GMT"}]}',
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "feed.json",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: '<?xml version="1.0" encoding="UTF-8"?>
<rss xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:wfw="http://wellformedweb.org/CommentAPI/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:sy="http://purl.org/rss/1.0/modules/syndication/" xmlns:slash="http://purl.org/rss/1.0/modules/slash/" xmlns:webfeeds="http://webfeeds.org/rss/1.0" version="2.0">
  <channel>
    <title>My RSS Feed</title>
    <link>https://example.com/</link>
    <atom:link href="https://example.com/feed.rss" rel="self" type="application/rss+xml"/>
    <lastBuildDate>Wed, 01 Jan 2020 00:00:00 GMT</lastBuildDate>
    <language>en</language>
    <generator>https://lume.land</generator>
    <author>
      <name>Laura Rubio</name>
    </author>
    <webfeeds:cover image="https://example.com/image.png"/>
    <webfeeds:logo>https://example.com/icon.svg</webfeeds:logo>
    <webfeeds:accentColor>#ff0000</webfeeds:accentColor>
    <item>
      <title>PAGE 5</title>
      <link>https://example.com/page5/</link>
      <guid isPermaLink="false">https://example.com/page5/</guid>
      <author>
        <name>Óscar</name>
        <uri>https://oscarotero.com</uri>
      </author>
      <content:encoded><![CDATA[Content of Page 5]]></content:encoded>
      <pubDate>Thu, 21 Jun 1979 23:45:00 GMT</pubDate>
      <atom:updated>1979-06-21T23:45:00.000Z</atom:updated>
    </item>
  </channel>
</rss>',
    data: {
      basename: "feed",
      content: '<?xml version="1.0" encoding="UTF-8"?>
<rss xmlns:content="http://purl.org/rss/1.0/modules/content/" xmlns:wfw="http://wellformedweb.org/CommentAPI/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:atom="http://www.w3.org/2005/Atom" xmlns:sy="http://purl.org/rss/1.0/modules/syndication/" xmlns:slash="http://purl.org/rss/1.0/modules/slash/" xmlns:webfeeds="http://webfeeds.org/rss/1.0" version="2.0">
  <channel>
    <title>My RSS Feed</title>
    <link>https://example.com/</link>
    <atom:link href="https://example.com/feed.rss" rel="self" type="application/rss+xml"/>
    <lastBuildDate>Wed, 01 Jan 2020 00:00:00 GMT</lastBuildDate>
    <language>en</language>
    <generator>https://lume.land</generator>
    <author>
      <name>Laura Rubio</name>
    </author>
    <webfeeds:cover image="https://example.com/image.png"/>
    <webfeeds:logo>https://example.com/icon.svg</webfeeds:logo>
    <webfeeds:accentColor>#ff0000</webfeeds:accentColor>
    <item>
      <title>PAGE 5</title>
      <link>https://example.com/page5/</link>
      <guid isPermaLink="false">https://example.com/page5/</guid>
      <author>
        <name>Óscar</name>
        <uri>https://oscarotero.com</uri>
      </author>
      <content:encoded><![CDATA[Content of Page 5]]></content:encoded>
      <pubDate>Thu, 21 Jun 1979 23:45:00 GMT</pubDate>
      <atom:updated>1979-06-21T23:45:00.000Z</atom:updated>
    </item>
  </channel>
</rss>',
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "feed.rss",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<p>This page is exported to <code>/404.html</code>, not <code>/404/index.html</code></p>
",
    data: {
      basename: "404",
      children: "<p>This page is exported to <code>/404.html</code>, not <code>/404/index.html</code></p>
",
      content: "This page is exported to \`/404.html\`, not \`/404/index.html\`
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/404.html",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/404",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<html><head></head><body>Content of Page 5</body></html>",
    data: {
      author: [
        "name",
        "url",
      ],
      basename: "page5",
      children: "Content of Page 5",
      content: "Content of Page 5",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      published: "1979-06-21T23:45:00.000Z",
      search: [],
      tags: "Array(0)",
      title: "Page **5**",
      url: "/page5/",
    },
    src: {
      asset: false,
      ext: ".yaml",
      path: "/page5",
      remote: undefined,
    },
  },
]
`;
